sklearn/neighbors/base.py

```diff
@@ -99,6 +98,55 @@ def _get_weights(dist, weights):
                          "'distance', or a callable function")
 
 
+def _decompose_neighbors_graph(graph, accept_unequal=True):
+    """Decompose a nearest neighbors sparse graph into distances and indices
+
+    Parameters
+    ----------
+    graph : CSR sparse matrix
+        Neighbors graph as given by kneighbors_graph or radius_neighbors_graph
+
+    accept_unequal : boolean, default=True
+        Whether to accept graphs with different number of neighbors for each
+        sample.
+
+    Returns
+    -------
+    distances : array, shape (n_samples, n_neighbors) or array of arrays
+        Distances to nearest neighbors.
+        It is an array of arrays if the number of neighbors is not identical
+        for every samples.
+
+    indices : array, shape (n_samples, n_neighbors) or array of arrays
+        Indices of nearest neighbors.
+        It is an array of arrays if the number of neighbors is not identical
+        for every samples.
+    """
+    n_samples = graph.shape[0]
+
+    # if there is the same number of neighbors for each sample
+    if np.unique(np.diff(graph.indptr)).size == 1:
+        distances = graph.data.reshape(n_samples, -1)
+        indices = graph.indices.reshape(n_samples, -1)
+        return distances, indices
+
+    else:  # radius neighbors case
+        if not accept_unequal:
+            raise ValueError(
+                'Precomputed sparse neighbors graph does not have the same '
+                'number of neighbors for each sample. Note that duplicated '
+                'samples should have an explicit zero entry in the sparse '
+                'neighbors graph.')
+
```

Comment:
We can have some fast/non-copy cases for kneighbors if we wish:
```
row_nnz = np.diff(X.indptr)
if row_nnz.max() == row_nnz.min():
    def extract(a):
        return a.reshape(n_samples, -1)[:, :n_neighbors]
else:
    idx = np.tile(np.arange(n_neighbors), n_samples).reshape(n_samples, n_neighbors)
    idx += np.indptr[:, np.newaxis]
    def extract(a):
        a.take(idx, mode='clip').reshape(n_samples, n_neighbors)
return extract(X.data), extract(X.indices)
```

Fast/non-copy paths with `radius_neighbors`:
```
if X.data.max() <= radius:
    return X.data, X.indices
else:
    X mask = X.data <= radius
    return np.compress(mask, X.data), np.compress(mask, X.indices)
```